db/channel_cache.go

```diff
@@ -173,6 +173,52 @@ func (c *channelCache) wouldBeImmediatelyPruned(change *LogEntry) bool {
 
 }
 
+// Remove purges the given doc IDs from the channel cache and returns the number of items removed.
+func (c *channelCache) Remove(docIDs []string, startTime time.Time) (count int) {
+	// Exit early if there's no work to do
+	if len(docIDs) == 0 {
+		return 0
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// Build subset of docIDs that we know are present in the cache
+	foundDocs := make(map[string]struct{}, 0)
+	for _, docID := range docIDs {
+		if _, found := c.cachedDocIDs[docID]; found {
+			foundDocs[docID] = struct{}{}
+		}
+	}
+
+	// Do the removals in one sweep of the channel cache
+	end := len(c.logs) - 1
+	for i := end; i >= 0; i-- {
+		if _, ok := foundDocs[c.logs[i].DocID]; ok {
+			docID := c.logs[i].DocID
+
+			// Make sure the document we're about to remove is older than the start time of the purge
+			// This is to ensure that resurrected documents do not accidentally get removed.
+			if c.logs[i].TimeReceived.After(startTime) {
+				base.Debugf(base.KeyCache, "Skipping removal of doc %q from cache %q - received after purge",
+					base.UD(docID), base.UD(c.channelName))
+				continue
+			}
+
+			// Memory-leak safe delete from SliceTricks:
```

Comment:
Validate this is safe for size 1 cache

Reply:
Tested manually in https://play.golang.org/p/qpQ82e8aCQ_C